### DISCUSSION TITLE
How do I get the properties of watcher in plugin developing

### DISCUSSION THREAD

**Discussion by WinterWaterWarm** [apache/skywalking#9958] [Q&A]
For example，plugin `trace-ignore-plugin` registers a `traceIgnorePatternWatcher` as BootService
`apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/resources/META-INF/services/org.apache.skywalking.apm.agent.core.boot.BootService`
![企业微信截图_16684138367783](https://user-images.githubusercontent.com/39685919/201609319-176b9ca0-6452-4745-aae9-d8b2cd857c6e.png)
`apm-sniffer/optional-plugins/trace-ignore-plugin/src/main/java/org/apache/skywalking/apm/plugin/trace/ignore/TraceIgnoreExtendService.java`
![企业微信截图_16684136855830](https://user-images.githubusercontent.com/39685919/201608876-5af8190e-5b10-4706-adef-d57aa8079a6c.png)
![企业微信截图_16684146808100](https://user-images.githubusercontent.com/39685919/201612519-9d929049-9145-4bfa-b165-4942f0b91a12.png)
How can I call the method `value` of watcher at any time in plugin developing?

**Comment by wu-sheng** [comment]
What your ask doesn't make any sense. You are developing something but don't aware of how it works?

**Reply by WinterWaterWarm** [reply]
Here's what I want to do:
1. Register a `AgentConfigWatcher` to collect dynamic configuraions named `plugin.http.include_http_headers`

```
import org.apache.skywalking.apm.agent.core.conf.dynamic.AgentConfigChangeWatcher;
import org.apache.skywalking.apm.agent.core.logging.api.ILog;
import org.apache.skywalking.apm.agent.core.logging.api.LogManager;

public class IncludeHttpHeaderWatcher extends AgentConfigChangeWatcher {
    private static final ILog LOGGER = LogManager.getLogger(IncludeHttpHeaderWatcher.class);

    private String includeHttpHeader;

    public IncludeHttpHeaderWatcher(final String propertyKey) {
        super(propertyKey);
        this.includeHttpHeader = getDefaultValue();
    }

    private void activeSetting(String config) {
        if (LOGGER.isDebugEnable()) {
            LOGGER.debug("Updating using new static config: {}", config);
        }

        try {
            this.includeHttpHeader = config;
        } catch (NumberFormatException ex) {
            LOGGER.error(ex, "Cannot load {} from: {}", getPropertyKey(), config);
        }
    }

    @Override
    public void notify(final ConfigChangeEvent value) {
        if (EventType.DELETE.equals(value.getEventType())) {
            activeSetting(getDefaultValue());
        } else {
            activeSetting(value.getNewValue());
        }
    }

    @Override
    public String value() {
        return includeHttpHeader;
    }

    private String getDefaultValue() {
        return "";
    }
}
```
i registed this watcher in BootService and it works.  I can set the dynamic `plugin.http.include_http_headers` to `includeHttpHeader`

```
@Override
    public void boot() {
        INCLUDE_HTTP_HEADER_WATCHER = new IncludeHttpHeaderWatcher("plugin.http.include_http_headers");
        ServiceManager.INSTANCE.findService(ConfigurationDiscoveryService.class)
                .registerAgentConfigChangeWatcher(INCLUDE_HTTP_HEADER_WATCHER);
    }
```

2. I want to develop a plugin to collect client header according to dynamic configuration `plugin.http.include_http_headers`.

so, how can i get the `watcher.includeHttpHeader` in my plugin code?

```

**Reply by WinterWaterWarm** [reply]
Is there something i didn't describe clearly，sincerely ask for help @wu-sheng

**Reply by wu-sheng** [reply]
You are developing your private features, I don't have time to support that.
As a member of community, I am only available to provide limited supports and feedback.

**Reply by WinterWaterWarm** [reply]
All right~
